Add "other permissions" to group permissions table

exler opened the pull request:
Resolves issue #5482 by replacing the "Other permissions" section with a new column in the "Object permissions" table.

Permissions not in the main categories (add, change, delete) will be placed in this column as an option in a multiple select box.

lb- reviewed (changes requested):
Hey @exler thanks for taking the time to submit this PR, code wise it is fine however user experience wise I think this may need some work.

## Feedback
Not sure on the best way forward but here is the feedback:
* The multi-select is a bit unintuitive as when you click something it goes grey (selected), but it is actually quite hard to see at a glance whether this grey is disabled/selected/unselected, especially when there are not many values listed.
* The text value says something like 'Can view Person' which makes sense when they were sitting in their own stand-alone list, however in the multi-select, this is quite verbose as you are already viewing the Person row.
* When even one permission exists, there is a large ~5 row select box shown.
* In general, I would be concerned that this UI will make it harder to manage permissions and not make mistakes, obviously it should be super clear what permissions are included or not.

## Suggestions
Again, no specific approach is recommended, just trying to get an understanding of what the problem is and how else we could solve this for CMS users and developers.

* It might be good to get a bit more clarity on what kind of problem is being solved here, do you have some mocks/screenshots of your scenarios (maybe with a really long permission per object list it is better with this new approach).
* What happens when there are zero other permissions for other objects, does the extra column still show? Maybe it can be removed if not applicable
* Instead of a multi-select, what about a multi-checkbox, this aligns with the other checkbox usage and is a lot harder to accidentally select/unselect something.
* What if this area was maybe just easier to customise, maybe including some `block` syntax so the template can be overridden in parts to allow for specific use cases to be presented.
* How do other UIs solve this, can we get some inspiration from them.

## Screenshots (reference)

The changes (Firefox) for the multi-select box look like this.

<img width="1097" alt="Screen Shot 2021-09-19 at 2 26 50 pm" src="https://user-images.githubusercontent.com/1396140/133915514-e1c288b8-9a29-4ce7-b26a-f72cb375b3ff.png">

The original looks like this.
![Screenshot 2021-09-19 at 14-27-45 Wagtail - Editing Editors](https://user-images.githubusercontent.com/1396140/133915521-0f78bed4-0538-41ab-8ba0-78e36e2a68bc.png)

exler commented:
Hello @lb-, thanks for the feedback!

I updated the PR to remove the model name from permission name to make it less verbose. I did not change the styles as I'm not quite sure what would be the proper design - there are libraries made with styled multiselect fields such as [Select2](https://select2.org/), but I'm not sure if it's worth adding just for such a simple feature. However, I don't think the default widget is confusing as there are only 2 states - selected/unselected.

The problem I'm solving here is the fact that permissions at this moment are unorganized. In the company I'm working we are using Wagtail and clients are constantly complaining that they have to search for the permissions where adding "Other permissions" as another column solves their problem. In the company fork we also added "View" as one of the primary columns (next to "Edit, Create, Delete"), but I decided not to add it due to [this comment](https://github.com/wagtail/wagtail/issues/5482#issuecomment-518201651). So the issue here is not a long list of permissions (which is also a problem because we frequently need to scroll through the whole page), but the ability to have all permissions for one model in one place.

I'm not sure how other UIs solve this as I'm not using any other CMS solutions, but I'm open to suggestions.

lb- commented on wagtail/users/templates/wagtailusers/groups/includes/formatted_permissions.html:
```diff
@@ -38,30 +40,17 @@ <h2>{% trans "Object permissions" %}</h2>
+                        {% endfor %}
+                    </select>
+                    {% endif %}
+                </td>
```
I have not tested this but it might be better to use a multi-checkbox as below (as per discussion, I agree that using select2 is overkill)

```
                <td>
                    {% if content_perms_dict.other %}
                    <fieldset class="other-permissions">
                        <legend>{% trans "Other permissions" %}</legend> 
                        {% for other_perm in content_perms_dict.other %}     
                        <label class="other-permissions-item">
                            <input type="checkbox" name="permissions" value="{{ other_perm.perm.id }}" {% if other_perm.selected %}checked{% endif %}>
                            {{ other_perm.name }}
                        </label>
                        {% endfor %}
                    </fieldset>
                    {% endif %}
                </td>
```

Also, I do not think the `other_perm.name` needs translation, I would assume that would already be provided as a translated value.

The `legend` can be hidden but should be there for screen readers.

As noted above, the `td` would also need to be not rendered if not needed.

lb- reviewed (changes requested):
A few more suggestions, main one though is the potential to use a set of checkboxes instead of multi-select.

exler commented:
Thanks for the review @lb-! 

I have added the proposed changes. The checkboxes look better, however I am concerned whether this will be readable with long lists of permissions. With 2-3 checkboxes it looks fine, but there will be a lot of whitespace if there are more permissions. Below I added the screenshot of the new changes:

![obraz](https://user-images.githubusercontent.com/9078240/134392198-e966cf06-d075-4c14-9732-a70f908728b1.png)

gasman commented:
Is there any way we can avoid the "Admin -> Can access Wagtail" permission being shown as an "object permission"? The fact that it's attached to a model is only an internal implementation detail - you can't define a permission without a content type, and a content type with no model creates other problems (see #6810).

exler commented:
Hey @gasman - sure we can. I implemented it just for this case, because I'm not sure if there will be other "other" permissions to put in the same section. This is how it looks now:

![obraz](https://user-images.githubusercontent.com/9078240/134401893-7197d572-d4c8-4d2c-a6ed-2c0655b2cfce.png)

lb- reviewed (changes requested):
@exler this is looking heaps better, thank you for taking the time to refine this solution.

Are you able to flesh out some additional unit tests for this?
* custom permissions exist / not exist
* a group that has some custom permissions checked (to test checked state)
* the other permissions grouping (probably already covered in unit tests)

I know this may take a bit of trial and error but adding unit tests for these new features is the best way to ensure that they do not break in the future.

exler commented:
Hey @lb-, I've looked into writing some tests for this PR, but I have no idea how can I programmatically register permissions to show up in the Groups view. Normally I add the permissions to `@hooks.register('register_permissions')` in the app's `wagtail_hooks.py` so that they show up as custom permissions, but I cannot do that from the test suite. Can you point me in the right direction?

exler commented:
Thanks for the help @lb-! I was able to write some simple tests to cover the cases you mentioned. 

Is there anything else I should add?

lb- commented on wagtail/users/static_src/wagtailusers/scss/groups_edit.scss:
```diff
@@ -16,6 +16,14 @@
     }
+
+    .custom-permissions {
+        padding-bottom: 0;
```
The default bold feels a bit off here, maybe let's  make that look the same as other content.

```suggestion
        padding-bottom: 0;
        
        label {
            font-weight: inherit;
        }
```

lb- commented on wagtail/users/static_src/wagtailusers/scss/groups_edit.scss:
```diff
@@ -16,6 +16,18 @@
+
+    .custom-permissions {
+        padding-bottom: 0;
+        
```
```suggestion

```

lb- commented:
@exler  sorry, could you please update your branch and fix the linting issue, I attempted to resolve in git but whitespace is causing issues.

This is looking really good, if all other tests pass we should be good to get this in soon.

@gasman any other thoughts on this one?

exler commented:
Sure thing @lb-, I updated the `.scss` file.

lb- commented:
@exler thanks heaps for your time on this, merged in via 8d42bec
